improver_tests/utilities/test_field_texture.py

```diff
@@ -0,0 +1,171 @@
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+""" Tests of FieldTexture plugin."""
+
+import iris
+import numpy as np
+import pytest
+
+from improver.synthetic_data.set_up_test_cubes import (
+    add_coordinate,
+    set_up_probability_cube,
+    set_up_variable_cube,
+)
+from improver.utilities.field_texture import FieldTexture
+
+NB_RADIUS = 10000
+RATIO_THRESH = 0.4
+
+
+@pytest.fixture(name="multi_cloud_cube")
+def multi_cloud_fixture():
+    """Set up a multi-realization cube containing cloud area fraction on an
+        equal area gride with 2km grid spacing."""
+
+    cloud_area_fraction = np.zeros((1, 10, 10), dtype=np.float32)
+    cloud_area_fraction[:, 1:4, 1:4] = 1.0
+    cube = set_up_probability_cube(
+        cloud_area_fraction,
+        variable_name="cloud_area_fraction",
+        threshold_units="1",
+        thresholds=[0.8125],
+        spatial_grid="equalarea",
+    )
+    cube = add_coordinate(cube, [0, 1, 2], "realization", dtype=np.int32)
+    cube = iris.util.squeeze(cube)
+    return cube
+
+
+@pytest.fixture(name="no_cloud_cube")
+def no_cloud_fixture():
+    """Set up a multi-realization cube containing no cloud present in the field
+        on an equal area grid with 2km grid spacing."""
+
+    cloud_area_fraction = np.zeros((1, 10, 10), dtype=np.float32)
+    cube = set_up_probability_cube(
+        cloud_area_fraction,
+        variable_name="cloud_area_fraction",
+        threshold_units="1",
+        thresholds=[0.8125],
+        spatial_grid="equalarea",
+    )
+    cube = add_coordinate(cube, [0, 1, 2], "realization", dtype=np.int32)
+    cube = iris.util.squeeze(cube)
+    return cube
+
+
+@pytest.fixture(name="all_cloud_cube")
+def all_cloud_fixture():
+    """Manipulate an existing multi-realization cube so the field is occupied
+       entirely by cloud."""
+
+    cloud_data = np.ones((3, 10, 10), dtype=np.float32)
+    cube = set_up_variable_cube(cloud_data, name="cloud_data", spatial_grid="equalarea")
+    return cube
```

Comment:
These functions are better, but there's some duplication and also the `all_cloud_fixture` is not a probability cube.  All three of your input cubes should have the same metadata and threshold coordinate, but different data.  So you can do this a bit more tidily.

Suggest rearranging your code slightly to do something like the following:
```
def cloud_probability_cube(data):
    cube = set_up_probability_cube(
        np.zeros((1, 10, 10), dtype=np.float32),
        variable_name="cloud_area_fraction",
        threshold_units="1",
        thresholds=[0.8125],
        spatial_grid="equalarea",
    )
    cube = add_coordinate(cube, [0, 1, 2], "realization", dtype=np.int32)
    cube = iris.util.squeeze(cube)
    cube.data = data
    return cube

@pytest.fixture(name="multi_cloud_cube")
def multi_cloud_fixture():
    cloud_area_fraction = np.zeros((3, 10, 10), dtype=np.float32)
    cloud_area_fraction[:, 1:4, 1:4] = 1.0
    return cloud_probability_cube(cloud_area_fraction)
```
And similar for your other two fixtures.

Reply:
Done.